packages/std/src/math.rs

```diff
@@ -247,49 +321,40 @@ impl fmt::Display for Uint128 {
 }
 
 impl ops::Add<Uint128> for Uint128 {
-    type Output = Self;
+    type Output = Result<Self, OverflowError>;
 
-    fn add(self, rhs: Self) -> Self {
-        Uint128(self.u128() + rhs.u128())
+    fn add(self, rhs: Self) -> Self::Output {
+        self.add(&rhs)
     }
 }
 
 impl<'a> ops::Add<&'a Uint128> for Uint128 {
-    type Output = Self;
-
-    fn add(self, rhs: &'a Uint128) -> Self {
-        Uint128(self.u128() + rhs.u128())
-    }
-}
-
-impl ops::AddAssign<Uint128> for Uint128 {
-    fn add_assign(&mut self, rhs: Uint128) {
-        self.0 += rhs.u128();
-    }
-}
+    type Output = Result<Self, OverflowError>;
 
-impl<'a> ops::AddAssign<&'a Uint128> for Uint128 {
-    fn add_assign(&mut self, rhs: &'a Uint128) {
-        self.0 += rhs.u128();
+    fn add(self, rhs: &'a Uint128) -> Self::Output {
+        let (op1, op2) = (self.u128(), rhs.u128());
+        op1.checked_add(op2)
```

Comment:
Why change the behavior here? 

 There is the `checked_add` operator (and more), I would keep the same public API of `+` (even if it is something like `self.checked_add(rhs).unwrap()` to make "panic on overflow" explicit)

Also, why calling `checked_add` on the `.u128()` and not the `Uint128` helper method.

Reply:
But why `Sub` is "checked" by default, but `Add` is panic by default?

Reply:
Because Sub will error in many normal cases (sending 13 tokens when I have 5), where Add will only error in some extreme case (entire ETH supply in wei can safely be added far from the Uint128 overflow)

Reply:
The initial motivation for this change is for the use case here(https://github.com/yihuang/cosmwasm-plus/blob/cw1155-impl/contracts/cw1155-base/src/contract.rs#L110).
When the `from` is `None`(which means mint new tokens), the user inputted `amount` get passed to the `+` operator, I thought it's a security vulnerability at first. If we always make sure it panic at overflow, then it's not a security vulnerability, but still I find it a little bit confused why it's different from the `Sub` operator.

Reply:
Let's keep the panic for add by default. This the behaviour of all Rust types as well and the only Reason why Uint128 exists is JSON serialization/deserialization of u128. It should be almost identical to u128.

To me the behaviour of `impl ops::Sub<Uint128> for Uint128` is what should get attention. Ideally we make it panic. However, I am a bit scared to just change the behaviour. This is why I suggested deprecating it.

Reply:
ok, I'll revert the `Add` change.
Do you want to put `deprecated` attribute on both `Add` and `Sub`?

Reply:
> ok, I'll revert the Add change.

Thanks!

> Do you want to put `deprecated` attribute on both `Add` and `Sub`? there's maybe also legit use case for that too? How about keep it as it for now?

What about this: `impl ops::Sub<Uint128> for Uint128` gets the following deprecate note: "The behaviour of the minus operator for Uint128 is likely to change in the future. Please consider using Uint128::checked_sub, Uint128::wrapping_sub or Uint128::saturating_sub."

Reply:
> When the from is None(which means mint new tokens), the user inputted amount get passed to the + operator directly, I thought it's a security vulnerability at first. If we always make sure it panic at overflow, then it's not a security vulnerability, but still I find it a little bit confused why it's different from the Sub operator.

Good point. And [relying on compiler flags](https://github.com/CosmWasm/cosmwasm/blob/main/contracts/hackatom/Cargo.toml#L22) for such security critical behavior is sub-optimal. 

I think a reimplementation using checked_add internally, but then panic on error (`.unwrap()`) will have the same external behavior but be very explicit and stable about how it handles overflow.

Reply:
> I think a reimplementation using checked_add internally, but then panic on error (.unwrap()) will have the same external behavior but be very explicit and stable about how it handles overflow.

Sounds good to me.

Reply:
> The behaviour of the minus operator for Uint128 is likely to change in the future. Please consider using Uint128::checked_sub, Uint128::wrapping_sub or Uint128::saturating_sub.

it seems unable to add deprecated annotation to trait impl: https://github.com/rust-lang/rust/issues/39935